### ISSUE TITLE
Code for forecasting task?

### ISSUE THREAD

**Issue by SaikiranKannaiah436** [argoai/argoverse-api#30]
Publishing the code for baseline models for forecasting task would be much appreciated.  Baseline results look really interesting and would push the state-of-the-art for map based prediction if the baseline models are available, especially the LSTM models.

**Comment by James-Hays** [comment]
Thanks for your feedback. We're working on it! We plan on releasing code for nearest neighbor baselines and trained models for the LSTM baselines [not necessarily code for the LSTM training, though].

The nearest neighbor baseline can work surprisingly well, actually, and it's easy to implement. You could start tinkering with that on your own. We definitely expect that are clever ways to use the map and outperform any of our baselines.

**Comment by l1xiao** [comment]
> Thanks for your feedback. We're working on it! We plan on releasing code for nearest neighbor baselines and trained models for the LSTM baselines [not necessarily code for the LSTM training, though].
> 
> The nearest neighbor baseline can work surprisingly well, actually, and it's easy to implement. You could start tinkering with that on your own. We definitely expect that are clever ways to use the map and outperform any of our baselines.

Does "nearest neighbor" mean the k-NN algorithm? Would you mind providing some references(website links or keywords) about nearest neighbor you mentioned? Thanks!

**Comment by jagjeet-singh** [comment]
Yes, we are referring to k-NN. The conventional Nearest Neighbor regressor (https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.KNeighborsRegressor.html) can do surprisingly well because of the large scale of the dataset. The lookup can be based on the observed 2 secs of the agent trajectory.

**Comment by Wilco17** [comment]
> Yes, we are referring to k-NN. The conventional Nearest Neighbor regressor (https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.KNeighborsRegressor.html) can do surprisingly well because of the large scale of the dataset. The lookup can be based on the observed 2 secs of the agent trajectory.

It's really interesting how a simple algorithm can perform so well (reported in your paper). As far as I know, the scikit-learn_ k-NN regressor does not work on multivariate-data. Have you implemented your custom bi-vraiate k-nn regressor?

**Comment by James-Hays** [comment]
Hi Wilco17, we haven't done anything fancy with our nearest neighbor baselines. I'm not sure if this is what you're referring to, but k-NN could have problems with multimodal data if you are averaging among the top matches. E.g. if you have an observed trajectory and 7 of its nearest neighbors go straight and 3 turn right, the average of those future trajectories is going to be something in between which is almost certainly implausible. But we're not averaging forecasts when we do k-NN. We're making and evaluating multiple forecasts. E.g. 10 guesses as to where the vehicle might go. We'll have a new version of the Argoverse paper soon on arXiv which is clearer about forecasting experiments.

**Comment by Wilco17** [comment]
Hi James-Hays, I was performing experiments just after my question and I rapidly come up with the same results and idea, nice! One of the main problems I encountered with this approach is the lack of robustness of the distance measurement between the 20 past seconds of the agent and their neighbors in the LUT due to the noise in the measurements. For instance, trajectories might be look dissimilar in the first 20 time-steps but similar in the following 30... I feel like that KNN coupled with some heuristics and distance measurements w.r.t to candidate lanes should work pretty well, but certainly intractable to be computed on-board on an embedded platform. 

Thank you very much for your contributions!!

**Comment by nnsriram97** [comment]
Hey thanks a lot for the dataset. Looking forward for your baseline codes. I would like to clarify on how you had normalized the trajectories for training. The paper mentions that the trajectory starts at origin for all the AGENTS and goes in the positive x direction. I would like to know how you had aligned these trajectories? Also, how is y at T_obs = 0? Kindly clarify. Thank you once again!

**Comment by johnwlambert** [comment]
@Wilco17 @SaikiranKannaiah436 @l1xiao Please find the code for the Argoverse forecasting baselines here: https://github.com/jagjeet-singh/argoverse-forecasting.

Our motion forecasting leaderboard is here: https://evalai.cloudcv.org/web/challenges/challenge-page/454/leaderboard/1279

**Comment by kargarisaac** [comment]
> @Wilco17 @SaikiranKannaiah436 @l1xiao Please find the code for the Argoverse forecasting baselines here: https://github.com/jagjeet-singh/argoverse-forecasting.
> 
> Our motion forecasting leaderboard is here: https://evalai.cloudcv.org/web/challenges/challenge-page/454/leaderboard/1279

Are any of these approaches open source? Is it possible to put the link of any of them wit the code in the leaderboard?

**Comment by James-Hays** [comment]
Hi Isaac,

It's up to the individual teams about whether they release source code. I'm not aware of any releases so far, but you could ask them. You can get more info about the winning techniques from our presentation at NeurIPS 2019 which is viewable here: 
https://slideslive.com/38923162/argoai-challenge
